z3/src/datatype_builder.rs

```diff
@@ -1,117 +1,179 @@
+use std::mem;
 use std::{convert::TryInto, ptr::null_mut};
 use z3_sys::*;
-use {Context, DatatypeBuilder, DatatypeSort, DatatypeVariant, FuncDecl, Sort, Symbol};
+use {
+    Context, DatatypeAccessor, DatatypeBuilder, DatatypeSort, DatatypeVariant, FuncDecl, Sort,
+    Symbol,
+};
 
 impl<'ctx> DatatypeBuilder<'ctx> {
-    pub fn new(ctx: &'ctx Context) -> Self {
+    pub fn new<S: Into<Symbol>>(ctx: &'ctx Context, name: S) -> Self {
         Self {
             ctx,
-            variants: Vec::new(),
+            name: name.into(),
+            constructors: Vec::new(),
         }
     }
 
-    pub fn variant(mut self, name: &str, fields: &[(&str, &Sort)]) -> Self {
-        let recognizer_name_sym = Symbol::String(format!("is-{}", name)).as_z3_symbol(self.ctx);
-        let name_sym = Symbol::String(name.to_string()).as_z3_symbol(self.ctx);
-
-        assert!(fields
-            .iter()
-            .all(|(name, sort)| sort.ctx.z3_ctx == self.ctx.z3_ctx));
-
-        let mut field_names: Vec<Z3_symbol> = Vec::with_capacity(fields.len());
-        let mut field_sorts = Vec::with_capacity(fields.len());
-
-        for (name, sort) in fields {
-            field_names.push(Symbol::String(name.to_string()).as_z3_symbol(self.ctx));
-            field_sorts.push(sort.z3_sort);
+    pub fn variant(&mut self, name: &str, fields: &'ctx [(&str, DatatypeAccessor<'ctx>)]) {
```

Comment:
Is there a reason this can't continue to be chainable by returning `&mut Self`?

Reply:
No, not really.